librad/src/git/util.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use git_ext::{is_not_found_err, reference};
+use std_ext::result::ResultExt as _;
+
+use super::{
+    refs::{self, Refs},
+    types::Namespace,
+};
+
+pub use super::{Storage, Urn};
+pub use git_ext::Tree;
+
+pub mod error {
+    use super::*;
+    use thiserror::Error;
+
+    #[derive(Debug, Error)]
+    #[non_exhaustive]
+    pub enum QuickCommit {
+        #[error("failed to update refs signature")]
+        Sigrefs(#[from] refs::stored::Error),
+
+        #[error(transparent)]
+        Git(#[from] git2::Error),
+    }
+}
+
+/// Quickly create a commit in the namespace and and on top of the branch
+/// described by [`Urn`], operating directly on [`Storage`].
+///
+/// This function is provided mainly for integration testing purposes, where the
+/// ceremony of a working copy is often undesired.
+///
+/// Note that the given [`Urn`]'s `path` will be sanitised such that it points
+/// to the locally-owned `refs/heads`. That is, `refs/remotes/xx/heads/pu` would
+/// be interpreted as `refs/namespaces/<urn>/refs/heads/xx/heads/pu`.
+///
+/// In other words, you can only use this function to commit to your own
+/// branches, but if you try to circumvent this restriction, the result may be
+/// surprising.
+///
+/// The default branch if the [`Urn`] contains no `path` is "master".
+#[tracing::instrument(
+    level = "debug",
+    skip(storage, urn, tree, message),
+    fields(urn = %urn),
+    err
+)]
+pub fn quick_commit(
+    storage: &Storage,
+    urn: &Urn,
+    tree: Tree,
+    message: &str,
+) -> Result<git2::Oid, error::QuickCommit> {
+    let repo = storage.as_raw();
+
+    let author = repo.signature()?;
+    let branch = {
+        let path = reference::Qualified::from(urn.path.clone().unwrap_or(reflike!("master")));
+        reflike!("refs/namespaces")
+            .join(Namespace::from(urn))
+            .join(reference::Qualified::from(reference::OneLevel::from(path)))
```

Comment:
Why do we qualify to go to one level and go back to qualified again? Seems a bit convoluted, or at least I'm confused by it.